### PR TITLE
Initial set of validation / test data for SAM

### PR DESCRIPTION
**Pull request by jkbonfield**
These have been lifted straight out the htslib's test harness, which in turn originally lifted them straight out of Staden io_lib's tests.

Some need combining or perhaps culling as they're more application testing rather than spec compliance testing.  It's a starting point at least for discussion.  I tried to add justification in the README file for why the files exist, but that also demonstrates how much we miss.  I'd like to be more precise, such as CIGAR_N.sam, CIGAR_P.sam, CIGAR_match.sam (M, =, X), CIGAR_indel.sam, FLAG_supp.sam, FLAG_secondary.sam, and so on.  It's hard to validate decoding worked unless we invent another text format (messy), but figured BAM to SAM testing at least does most of what we want with regards to interpretation.  Maybe just some text comments in the headers are sufficient?

I also wonder if we want to separate into valid (should accept), invalid (ideally should reject), and "inconsistent but legal format" (eg TLEN or MD being computed wrongly).

Maybe we could also define compliance categories for each format.  For example a CRAM decoder could be considered as basic if it handles the defaults of 1 slice per container, external references only, gzip only;  medium if it handles multi-slice containers, embedded references too; full if it also has lzma and bzip2 support.  All those up for debate of course.  The reason is it would be nice to be able to list implementations and at least say it passes compliance tests to level X rather than simply "FAIL" purely on the grounds of lacking lzma support.

Note I'm not proposing tests for CRAM v2.1 and earlier.

We also need considerably more CRAM files to test specific things in the decoder, eg
- Multi-ref slices
- Multiple slices per container
- Extremities of sizes; eg 1 seq per slice
- External vs embedded vs no-ref tests
- Everything in one CORE block (forcing the order of decode operations to be validated),
- Different compression codecs
- Corner case values (e.g. boundaries of ITF8 / LTF8)

### CODE REVIEW

File: test/SAM/README.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	Record "Fred" is integers, floats and strings.
+	Record "Jim" is byte-arrays and hex values.
+
+c1#bounds.sam
+realn01.sam
+realn02.sam
+	Contains sequence with alignments that extend beyond the end
+	of the reference.
+
+	Is this valid?  Code should at least cope, wither work or
```
Comment (yfarjoun) [inline test/SAM/README.md:32]:
```suggestion
	Code should cope, either work or
```

**Comment by yfarjoun** [comment]
Could we _not_ use `#` in filenames? While legal it can make it difficult to put into scripts etc. some software (cough cough, GATK) seems to have problems with that particular character in filenames...

File: test/SAM/ce#tag_padded.sam [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+III	0	CHROMOSOME_I	2	1	101M	*	0	0	CCTAGCCCTAACCCTAACCCTAACCCTCAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAA	###########################"##@B?8B?BA@@DDBCDDCBC@CDCDCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC	PT:Z:1;1;+;COMM|101;101;+;COMM
+IV	16	CHROMOSOME_I	2	1	27M1D73M	*	0	0	CCTAGCCCTAACCCTAACCCTAACCCTAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAA	#############################@B?8B?BA@@DDBCDDCBC@CDCDCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
+V	0	CHROMOSOME_I	2	1	27M1D73M	*	0	0	CCTAGCCCTAACCCTAACCCTAACCCTAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAA	#############################@B?8B?BA@@DDBCDDCBC@CDCDCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC
+*	768	CHROMOSOME_I	2	255	1M	*	0	0	*	*	CT:Z:+;STOP
+*	768	CHROMOSOME_I	28	255	3M	*	0	0	*	*	CT:Z:+;COMM;Note=consensus%0Amulti%09line%0Atag%0A!"#$%25^&*()_+<>?:@~{}%7C%0A1234567890-=[]'%3B#,./\
```
Comment (yfarjoun) [inline test/SAM/ce#tag_padded.sam:10]: since the spec calls out sepcifically how this CT tag should be interpreted, should we have a test for that? meaning, should we have a file that contains the GFF equivalent version of the tag and perhaps also the decoded version of the string text?

File: test/SAM/README.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	is not spec compilant.
+
+xx#triplet.sam
+	A template with 3 components (READ1, READ2 and neither).
+
+xx#unsorted.sam
+	Neither name nor chr / position sorted.
+
+
+index.sam	?
```
Comment (yfarjoun) [inline test/SAM/README.md:152]: 1. should the bams and crams include an index so that implementations can check that they can all read & use one agreed-upon index implementation?
1. in a similar vein, since we seem to have many different index naming scenes allowed, should we have some tests for that:
* file.bam + file.bai 
* file.bam + file.bam.bai
* file.sam + file.sai ?
* file.sam + file.sam.sai?
* ....

3. what about search order, i.e. in the presence of two indexes, file.bam.bai & file.bai, do we want to specify which index should be used?

File: test/SAM/xx#tlen.sam [hunk 4]
```diff
@@ -0,0 +1,34 @@
+@CO	xx has been encoded using the SAM spec; leftmost to rightmost
```
Comment (yfarjoun) [inline test/SAM/xx#tlen.sam:1]: SAM spec is now agnostic :-)

**Comment by yfarjoun** [comment]
- htsjdk looks for a .dict file adjacent to fa or fast files...are they standard? should they be?
- fasta files can have many extensions (fa, fasta, fa.gz, fasta.gz) should we have tests with all of them?
-

**Comment by jkbonfield** [comment]
I went through and rewrote all my SAM tests.  The old test files from htslib have been punted sideways to SAM.WIP so I have something to crib from for additional tests if needed, but in the fullness of time it will be culled.  It still needs SAM header tests adding so I probably won't tidy that up until we've got a complete test set.  Comments welcomed on what's there right now.

I've broken the tests into pass (must be accepted and ideally should never produce warnings), fail (*may* fail, depending on how anal you want your implementation to be), and warn (files which are valid SAM format but we may want to consider warning about).  Examples of warnings could be lowercase letters in SAM or even sequence like "Z", which is permitted in the spec but it's never going to work for BAM so is generally inadvisable.  Also in this category are tests for the various assumptions, like CIGAR has no meaning for unmapped data.  That technically means it's valid to have one, but again it's also appropriate and IMO advisable, to warn about it.

I also decided this PR will be SAM only, with their own PRs for BAM and CRAM.  I haven't included anything on specific tag types either, only the general tag format itself.  I think that's correct, as we can have per-tag specific test directories (eg as in the base modification one).

I've started on CRAM too: my branch for this is https://github.com/jkbonfield/hts-specs/tree/CRAM_tests.  My aim there is to provide test files that also potentially act as a narrative for writing a CRAM decoder.  Hence they can be processed in order with none of the earlier files requiring knowledge of how to decode the more complex topics included in the later files.  Some of this is requiring local hacks of the code too to generate the data, so it's slow progress.

**Comment by jkbonfield** [comment]
Revamped it.  Chucked out the old htslib test data and it's now all new tests per field in the header and per column in SAM.
The directories are now lowercase too.

Gone are BAM / CRAM tests.  These will be the subject of another PR (although likely not mine for BAM as I'll be concentrating on CRAM for some time).

There are no auxiliary tag tests either, as that should also be done as separate PR(s) IMO given it is a separate specification document.

**Comment by cmdcolin** [comment]
Is there any file that has a CG tag example by chance?

**Comment by jkbonfield** [comment]
No, as it's labelled as BAM only (although it can turn up in SAM in cases of older BAM to SAM conversion tools).

We don't yet have any BAM tests here.  I was leaving that to others, having been somewhat burnt out doing both SAM and CRAM. :-)

The easiest way to get it though is to create an artificial sequence with > 65535 cigar operations (eg 1M1D1M1D repeated) and convert to BAM using samtools.

**Comment by cmdcolin** [comment]
@jkbonfield thanks forgot that the CG tag is not actually in the SAM. that was enough to get me goin

**Review by tskir — APPROVED** [review]

**Comment by tskir** [comment]
As discussed on the File Formats call 2021-04-01 between @jmarshall, @lbergelson, @tcezard and @tskir, this PR **will be merged in its current form** in four weeks (2021-04-29), unless there are strong objections.

**Comment by jmarshall** [comment]
Suggest renaming _test/sam/SAM.md_ to _test/sam/README.md_ so it appears in the repository view when you click down to that directory.

**Comment by tskir** [comment]
@jmarshall  Good point, addressed :)
